### PR TITLE
Fixing headergroup bugs

### PR DESCRIPTION
**Pull request by catalinaperalta**
This PR updates the way that byte arrays are parsed to strings in the generated code and also fixes the value that's passed in as the header key when setting specific request headers, the generator now uses the original value passed in through the swagger. 

Also updated the corresponding headergroup tests and adding generated code. 

The issue tracking pending fixes is: #270

### CODE REVIEW

File: test/autorest/generated/headergroup/internal/headergroup/header.go [hunk 1]
```diff
@@ -313,7 +318,11 @@ func (HeaderOperations) ResponseDateHandleResponse(resp *azcore.Response) (*Head
 	if !resp.HasStatusCode(http.StatusOK) {
 		return nil, newError(resp)
 	}
-	return &HeaderResponseDateResponse{StatusCode: resp.StatusCode}, nil
+	val, err := time.Parse(time.RFC3339, resp.Header.Get("value"))
```
Comment (catalinaperalta) [inline test/autorest/generated/headergroup/internal/headergroup/header.go]: This isn't working properly yet. The issue is tracked in #270. I'm taking a look at it.

File: test/autorest/generated/headergroup/internal/headergroup/header.go [hunk 2]
```diff
@@ -329,7 +338,11 @@ func (HeaderOperations) ResponseDatetimeHandleResponse(resp *azcore.Response) (*
 	if !resp.HasStatusCode(http.StatusOK) {
 		return nil, newError(resp)
 	}
-	return &HeaderResponseDatetimeResponse{StatusCode: resp.StatusCode}, nil
+	val, err := time.Parse(time.RFC3339, resp.Header.Get("value"))
```
Comment (catalinaperalta) [inline test/autorest/generated/headergroup/internal/headergroup/header.go]: Same parse issue

File: test/autorest/generated/headergroup/internal/headergroup/header.go [hunk 3]
```diff
@@ -345,7 +358,11 @@ func (HeaderOperations) ResponseDatetimeRFC1123HandleResponse(resp *azcore.Respo
 	if !resp.HasStatusCode(http.StatusOK) {
 		return nil, newError(resp)
 	}
-	return &HeaderResponseDatetimeRFC1123Response{StatusCode: resp.StatusCode}, nil
+	val, err := time.Parse(time.RFC3339, resp.Header.Get("value"))
```
Comment (catalinaperalta) [inline test/autorest/generated/headergroup/internal/headergroup/header.go]: Same parse issue.

Comment (catalinaperalta) [reply test/autorest/generated/headergroup/internal/headergroup/header.go]: It compiles it just doesnt know how to respond to data returned from the endpoint.

File: test/autorest/generated/headergroup/internal/headergroup/header.go [hunk 4]
```diff
@@ -377,7 +398,11 @@ func (HeaderOperations) ResponseDurationHandleResponse(resp *azcore.Response) (*
 	if !resp.HasStatusCode(http.StatusOK) {
 		return nil, newError(resp)
 	}
-	return &HeaderResponseDurationResponse{StatusCode: resp.StatusCode}, nil
+	val, err := time.ParseDuration(resp.Header.Get("value"))
```
Comment (catalinaperalta) [inline test/autorest/generated/headergroup/internal/headergroup/header.go]: Same parse issue.

File: src/generator/protocol/operations.ts [hunk 5]
```diff
@@ -196,15 +285,15 @@ function createProtocolRequest(client: string, op: Operation, imports: ImportMan
     // add specific request headers
     const headerParam = values(op.request.parameters).where((each: Parameter) => { return each.protocol.http!.in === 'header'; });
     headerParam.forEach(header => {
-      text += `\treq.Header.Set("${header.language.go!.name}", ${formatParamValue(header, imports)})\n`;
+      text += `\treq.Header.Set("${header.language.default.name}", ${formatParamValue(header, imports)})\n`;
```
Comment (jhendrixMSFT) [inline src/generator/protocol/operations.ts]: Might want to add a comment why the default name is used here.

File: test/autorest/generated/headergroup/internal/headergroup/header.go (hunk 3 above)
Comment (jhendrixMSFT) [reply test/autorest/generated/headergroup/internal/headergroup/header.go]: Try `time.RFC1123Z`, or possibly the variant without the Z suffix.

File: src/generator/protocol/operations.ts [hunk 6]
```diff
@@ -148,6 +154,89 @@ function formatParamValue(param: Parameter, imports: ImportManager): string {
...
+  let text = ``;
+  switch (header.schema.type) {
+    case SchemaType.Boolean:
+      imports.add('strconv');
+      text = `\tval, err := strconv.ParseBool(resp.Header.Get("${header.header}"))\n`;
+      text += `\tif err != nil {\n`;
+      text += `\t\treturn nil, err\n`;
+      text += `\t}\n`;
+      headerText.body = text;
+      headerText.respObj = `{RawResponse: resp.Response, ${header.name}: &val}`;
```
Comment (jhendrixMSFT) [inline src/generator/protocol/operations.ts]: The "RawResponse: resp.Response" part is declared by the caller.  I suggest making this part a parameter to this function so it's only defined once by the caller.

File: test/autorest/generated/headergroup/internal/headergroup/header.go (hunk 3 above)
Comment (catalinaperalta) [reply test/autorest/generated/headergroup/internal/headergroup/header.go]: Yup I just figured out the fix, but I think I'll leave it for the next PR where I fix all the time/date related issues.

File: src/generator/protocol/operations.ts (hunk 5 above)
Comment (catalinaperalta) [reply src/generator/protocol/operations.ts]: Done

File: src/generator/protocol/operations.ts (hunk 6 above)
Comment (catalinaperalta) [reply src/generator/protocol/operations.ts]: Ok added it and passed it into the function as a param

File: test/autorest/generated/headergroup/internal/headergroup/models.go [hunk 7]
```diff
@@ -39,175 +41,189 @@ func (e Error) Error() string {
...
 }
 
 // HeaderResponseExistingKeyResponse contains the response from method Header.ResponseExistingKey.
 type HeaderResponseExistingKeyResponse struct {
-	// StatusCode contains the HTTP status code.
-	StatusCode int
+	// RawResponse contains the underlying HTTP response.
+	RawResponse *http.Response
+	UserAgent *string
 }
```
Comment (jhendrixMSFT) [inline test/autorest/generated/headergroup/internal/headergroup/models.go]: You will also want to add any description text as a comment above the header value.  If the description is empty we can put some more generic text like "UserAgent contains the value from the User-Agent header response." or something like that.

Comment (jhendrixMSFT) [reply test/autorest/generated/headergroup/internal/headergroup/models.go]: And include a new-line between entries.

Comment (catalinaperalta) [reply test/autorest/generated/headergroup/internal/headergroup/models.go]: Ok done.

**Pull request merged**